File: src/main/java/org/tdf/lotusvm/types/DataSection.java
package org.tdf.lotusvm.types;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.tdf.lotusvm.common.BytesReader;
import org.tdf.lotusvm.common.Vector;

import java.util.ArrayList;
import java.util.List;

public class DataSection extends AbstractSection {
    @Getter
    private List<DataSegment> dataSegments;

    public DataSection(SectionID id, long size, BytesReader payload, int offset, int limit) {
        super(id, size, payload, offset, limit);
    }

    @Override
    void readPayload() {
        dataSegments = DataSegment.readDataSegmentsFrom(getReader());
    }

    @Getter
    @AllArgsConstructor(access = AccessLevel.PRIVATE)
    public static class DataSegment {
        private final int memoryIndex;

        private final Instruction[] expression;

        private final byte[] init;

        public static DataSegment readFrom(BytesReader reader) {
            return new DataSegment(reader.readVarUint32(),
                Instruction.readExpressionFrom(reader),
                Vector.readBytesFrom(reader)
            );
        }

        public static List<DataSegment> readDataSegmentsFrom(BytesReader reader) {
            int length = reader.readVarUint32();
            List<DataSegment> res = new ArrayList<>(length);
            for (int i = 0; i < length; i++) {
                res.add(readFrom(reader));
            }
            return res;
        }
    }
}

File: src/main/java/org/tdf/lotusvm/types/DataSection.kt
package org.tdf.lotusvm.types

import org.tdf.lotusvm.common.BytesReader
import org.tdf.lotusvm.common.ObjectReader

class DataSection(id: SectionID, size: Long, payload: BytesReader, offset: Int, limit: Int) :
    AbstractSection(id, size, payload, offset, limit) {
    var dataSegments: List<DataSegment> = emptyList()
        private set

    public override fun readPayload() {
        dataSegments = reader.readObjectVec(DataSegment.Companion)
    }

    data class DataSegment(val memoryIndex: Int, val expression: Long, val init: ByteArray) {
        companion object : ObjectReader<DataSegment> {
            override fun readFrom(reader: BytesReader): DataSegment {
                return DataSegment(
                    reader.readVarUint32(),
                    reader.insPool.readExpressionFrom(reader),
                    reader.readByteVec()
                )
            }
        }
    }
}